twitcher/owssecurity.py

```diff
@@ -53,20 +63,85 @@ def prepare_headers(request, access_token):
                 LOGGER.debug("Prepared request headers.")
         return request
 
-    @staticmethod
-    def check_request(request):
+    def verify_access(self, request, service):
+        # TODO: public service access handling is confusing.
+        try:
+            if service.auth == 'cert':
+                verify_cert(request)
+            else:  # token
+                self._verify_access_token(request)
+        except OWSAccessForbidden:
+            if not service.public:
+                raise
+
+    def _verify_access_token(self, request):
+        try:
+            # try to get access_token ... if no access restrictions then don't complain.
+            token = self.get_token_param(request)
+            access_token = self.tokenstore.fetch_by_token(token)
+            if access_token.is_expired():
+                raise OWSAccessForbidden("Access token is expired.")
+            # update request with data from access token
+            # request.environ.update(access_token.data)
+            # TODO: is this realy the way we want to do this?
+            request = self.prepare_headers(request, access_token)
+        except AccessTokenNotFound:
+            raise OWSAccessForbidden("Access token is required to access this service.")
+
+    def check_request(self, request):
         protected_path = request.registry.settings['twitcher.ows_proxy_protected_path']
         if request.path.startswith(protected_path):
-            service_name = parse_service_name(request.path, protected_path)
+            # TODO: refactor this code
+            try:
+                service_name = parse_service_name(request.path, protected_path)
+                service = self.servicestore.fetch_by_name(service_name)
+                if service.public is True:
+                    LOGGER.warn('public access for service %s', service_name)
+            except ServiceNotFound:
+                # TODO: why not raising an exception?
+                service = Service(url='unregistered', public=False, auth='token')
+                LOGGER.warn("Service not registered.")
+            ows_request = OWSRequest(request)
+            if not ows_request.service_allowed():
+                raise OWSInvalidParameterValue(
+                    "service %s not supported" % ows_request.service, value="service")
+            if not ows_request.public_access():
+                self.verify_access(request, service)
+
+
+try:
+    from magpie.services import service_factory
+    from magpie.models import Service
+    from magpie import evaluate_call, verify_param
+except ImportError:
+    LOGGER.warning('Can not import magpie')
+
+from pyramid.httpexceptions import HTTPForbidden, HTTPNotFound
+from pyramid.interfaces import IAuthenticationPolicy, IAuthorizationPolicy
```

Comment:
All imports should be moved to the top with other imports.

Reply:
Done that way only to avoid merge errors... In fact the OWSSecurityMagpie should be done in its own file... will be removed in the magpie auth refactoring